File: 2021/7/python/main.py
import numpy as np

with open('../input.txt') as f:
    a = np.array([int(x) for x in f.read().strip().split(',')])

# Part 1
print("Part 1:", np.min([np.abs(a - i).sum() for i in range(999)]))

# Part 2
def f(i):
    return i*(i + 1)//2

print("Part 2:", np.min([f(np.abs(a - i)).sum() for i in range(999)]))

File: 2021/7/rust/src/main.rs
use std::error::Error;
use std::fs::File;
use std::io::{BufRead, BufReader};

fn main() -> Result<(), Box<dyn Error>> {
    let file = File::open("../input.txt")?;
    let mut reader = BufReader::new(file);
    let mut line = String::new();
    reader.read_line(&mut line)?;
    let numbers: Vec<i64> = line
        .trim()
        .split(",")
        .into_iter()
        .map(|n| n.to_string().parse::<i64>().unwrap())
        .collect();

    // Part 1
    let r = (1..999)
        .map(|i| numbers.iter().map(|n| (n - i).abs()).sum::<i64>())
        .into_iter()
        .min();
    println!("Part 1: {}", r.unwrap());

    // Part 2
    let r = (1..999)
        .map(|i| {
            numbers
                .iter()
                .map(|n| {
                    let x = (n - i).abs();
                    x * (x + 1) / 2
                })
                .sum::<i64>()
        })
        .into_iter()
        .min();
    println!("Part 2: {}", r.unwrap());

    Ok(())
}
